snuba/datasets/generic_metrics_processor.py

```diff
@@ -0,0 +1,130 @@
+import logging
+import zlib
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Mapping, MutableMapping, Optional, Sequence, Tuple
+
+from snuba.consumers.types import KafkaMessageMetadata
+from snuba.datasets.events_format import EventTooOld, enforce_retention
+from snuba.datasets.metrics_messages import is_set_message, values_for_set_message
+from snuba.processor import (
+    InsertBatch,
+    MessageProcessor,
+    ProcessedMessage,
+    _ensure_valid_date,
+)
+
+logger = logging.getLogger(__name__)
+
+# These are the hardcoded values from the materialized view
+GRANULARITY_ONE_MINUTE = 1
+GRANULARITY_ONE_HOUR = 2
+GRANULARITY_ONE_DAY = 3
+
+
+class GenericMetricsBucketProcessor(MessageProcessor, ABC):
+    @abstractmethod
+    def _should_process(self, message: Mapping[str, Any]) -> bool:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _process_values(self, message: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    #
+    # This is mainly split out from _hash_timeseries_id for unit-testing purposes
+    #
+    def _timeseries_id_token(self, message: Mapping[str, Any]) -> bytearray:
+        field_separator: int = ord(";")
+        kv_separator: int = ord(",")
+        kv_assign: int = ord("=")
+        org_id: int = message["org_id"]
+        project_id: int = message["project_id"]
+        metric_id: int = message["metric_id"]
+        sorted_tag_items: Sequence[Tuple[str, int]] = sorted(message["tags"].items())
+
+        buffer = bytearray()
+        for field in [org_id, project_id, metric_id]:
+            buffer.extend(field.to_bytes(length=4, byteorder="little"))
+            buffer.append(field_separator)
```

Comment:
Up to you, but this is an opaque identifier, The separators do nothing useful. You are not trying to use it to establish uniqueness, in such case you would not want ambiguity. 
As long as the same time series always yields the same id and they are reasonably evenly distributed you are fine.